commander.go

```diff
@@ -0,0 +1,115 @@
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"io/fs"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+)
+
+// commander keeps track of a directory and runs commands on them providing
+// a crude form of parallelism. Not safe for concurrent use.
+type commander struct {
+	// path is guaranteed to be the absolute current path of commander.
+	path string
+	// checkin is a buffered channel. It's length limits the amount of goroutines running commands at once.
+	checkin chan struct{}
+}
+
+func newCommander(goroutines int) commander {
+	path, err := os.Getwd()
+	if err != nil {
+		log.Fatal("commander: getting current dir:", err)
+	}
+	if goroutines < 1 {
+		log.Fatal("commander: invalid number of goroutines argument")
+	}
+	log.Println("setting max simultaneous commands to", goroutines)
+	return commander{
+		path:    path,
+		checkin: make(chan struct{}, goroutines),
+	}
+}
+
+func (c *commander) cloneOrUpdateRepo(repo string) {
+	if _, err := c.Stat(repo); err != nil {
+		// Repo does not exist.
+		log.Printf("repo not found. cloning %s", repo)
+		d := filepath.Dir(repo)
+		if _, err := c.Stat(repo); err != nil {
+			log.Printf("creating directory %s", d)
+			c.Mkdir(d, dirMode)
+		}
+		c.Chdir(d)
+		c.Run("git", "clone", fmt.Sprintf("%s/%s", hostURL, repo))
+		return
+	}
+
+	c.Chdir(repo)
+	log.Printf("repo exists, updating %s", repo)
+	c.Run("git", "fetch")
+	c.Run("git", "pull")
+}
+
+func (r commander) Stat(path string) (os.FileInfo, error) {
+	return os.Stat(r.join(path))
+}
+func (r commander) Mkdir(d string, mode fs.FileMode) error {
+	return os.MkdirAll(r.join(d), mode)
+}
+func (r *commander) Chdir(path string) {
+	r.path = r.join(path)
+}
+func (r commander) join(path string) string {
+	if filepath.IsAbs(path) {
+		return path
+	}
+	return filepath.Join(r.path, path)
+}
+
+// Start begins command execution in commander's current directory and returns immediately.
+// Prints command output on finish. Calls os.Exit(1) on error.
+func (r *commander) Start(name string, arg ...string) {
+	r.run(true, name, arg...)
+}
+
+// Run executes command in commander's current directory and waits for it to finish.
+// Prints command output. If command returns non-zero exit code then result is logged
+// and os.Exit(1) is called.
+func (r *commander) Run(name string, arg ...string) {
+	r.run(false, name, arg...)
+}
+
+func (r *commander) run(async bool, name string, arg ...string) {
+	r.checkin <- struct{}{} // Check-in for work.
+	cmd := exec.Command(name, arg...)
+
+	var b bytes.Buffer
+	cmd.Stdout = &b
+	cmd.Stderr = &b
+	cmd.Dir = r.path
```

Comment:
I'm concerned with all the async and concurrent command executions, this `r.path` might be pointing at the wrong location for this particular command.

Reply:
So `commander` is not really safe for concurrent use (as in `go commander.Run`). For our use case this is OK, we have one commander which manages all commands from `main()`, at no point do we call a goroutine from `main()`, thus leaving all the concurrency for `commander` to manage. 

This might not be very idiomatic as an API since it's usually preferable to leave concurrency to the user. I'll bring forward the following counterpoint:  _bringing in concurrency into `main` will take a toll on readability. Right now `main` benefits from how `commander` abstracts away the concurrency_. For now we can make use of `commander` to make our lives easier and instantiate more `commander`s if we see we need to bring in concurrency into the main thread.

Reply:
And hey, even the Go standard library does not abide by this concurrency principle, you need only look at `exec.Command`'s `Start()` method, which does pretty much the same thing `commander.Start()` does. (this was no accident :smile:)